File: alien_wallpaper.rb
#!/usr/bin/env ruby

# Name: alien_wallpaper.rb
# Description: Downloads top images from specified subreddits.
# Author: Bob Gardner
# Updated: 4/18/16
# License: MIT

require 'json'
require 'net/http'
require 'open-uri'
require 'optparse'

# Number of images to download per subreddit/multireddit.
NUM_IMAGES = 25

# The default subreddits to download images from.
DEFAULT_SUBREDDITS = ['ArchitecturePorn',
                      'CityPorn',
                      'EarthPorn',
                      'SkyPorn',
                      'spaceporn',
                      'winterporn',
                      'quoteporn']

# Supported file types.
FILE_EXTENSION_REGEX = /\.(jpg|jpeg|gif|png)$/i

# The filename to save the post's image as.
#   precondition: post's image is a supported filetype.
def filename(post)
  post['data']['id'] + post['data']['url'].match(FILE_EXTENSION_REGEX)[0]
end

# Ignore nil, self posts, and unsupported images.
def ignore_post?(post)
  return true if post.nil?
  return true if post['data']['is_self']
  return false if post['data']['url'].match(FILE_EXTENSION_REGEX)
  $stderr.puts post['data']['url']
  true
end

# Download the image associated with the Reddit post.
def download_image(post, output_dir)
  # Construct the filename and filepath
  filepath = File.join(output_dir, filename(post))
  puts filepath

  # Need to use write-binary (wb) mode for images
  File.open(filepath, 'wb') { |f| f << open(post['data']['url']).read }
rescue SocketError
    abort('No Internet connection available')
rescue OpenURI::HTTPError => error
    $stderr.puts error
rescue OpenSSL::SSL::SSLError => error
    $stderr.puts error
end

# Return the json representation of `subreddit`
def fetch_subreddit_data(url, count = 25, after = '')
  params = "count=#{count}&after=#{after}"
  JSON.parse(URI.parse("#{url}.json?#{params}").read)
rescue SocketError
  abort('No Internet connection available')
end

# Download `n` images from `url` to `out_dir`
def download_images_for_url(url, n, out_dir)
  json = fetch_subreddit_data(url)
  images_remaining = n
  catch(:done) do
    loop do
      json['data']['children'].each do |post|
        throw :done if images_remaining <= 0
        download_image(post, out_dir) unless ignore_post?(post)
        images_remaining -= 1
      end
      json = fetch_subreddit_data(url, json['data']['after'])
    end
  end
end

def multi_to_url(multi)
  user, multi = multi.split('/')
  "https://www.reddit.com/user/#{user}/m/#{multi}"
end

def subreddit_to_url(subreddit)
  "https://www.reddit.com/r/#{subreddit}"
end

# The command line options.
def cli_options
  options = { out: Dir.pwd, s: DEFAULT_SUBREDDITS }
  OptionParser.new do |opts|
    opts.on('--subreddits=x,y,z', Array, 'List of subreddits') do |s|
      options[:s] = s
    end
    opts.on('--multi M', String, 'user/multireddit') do |m|
      options[:m] = m
    end
    opts.on('--out DIR', String, 'Directory to save wallpaper to') do |out|
      options[:out] = out
    end
    opts.on_tail('-h', '--help', 'Show this message') do
      puts 'Download beautiful wallpaper from subreddits'
      puts opts
      exit
    end
  end.parse!
  options
end

def main(opts)
  threads = []
  urls = opts[:s].map { |s| subreddit_to_url(s) }
  urls << multi_to_url(opts[:m]) unless opts[:m].nil?
  urls.each do |url|
    threads << Thread.new do
      download_images_for_url(url, NUM_IMAGES, opts[:out])
    end
  end
  threads.each(&:join)
end

main(cli_options)

File: alien_wallpaper.py
#!/usr/bin/env python3
"""
Download images from Reddit.

```bash
$ alien_wallpaper --help
```

"""

import argparse
import json
import logging
import multiprocessing
import os.path
import re
import shutil
import sys
import typing
import urllib.parse
import urllib.request

IMAGES_PER_SUBREDDIT = 25
DEFAULT_SUBREDDITS = ['ArchitecturePorn', 'CityPorn', 'EarthPorn', 'SkyPorn',
                      'spaceporn', 'winterporn', 'quoteporn']
USER_AGENT = 'AlienWallpaper by /u/WolfBlackout'

FILE_EXTENSION_PROG = re.compile('.*.(jpg|jpeg|gif|png)$', re.IGNORECASE)
Url = typing.NewType('Url', str)


class Post:
    def __init__(self, post_id: str, is_self: bool, url: Url) -> None:
        self.id = post_id
        self.is_self = is_self
        self.url = url

    @property
    def is_valid(self) -> bool:
        return not self.is_self and bool(FILE_EXTENSION_PROG.match(self.url))

    @property
    def filename(self) -> typing.Optional[str]:
        if self.extension is not None:
            return self.id + '.' + self.extension

    @property
    def extension(self) -> typing.Optional[str]:
        match = FILE_EXTENSION_PROG.match(self.url)
        if match is not None:
            return match.group(1)

    def download(self, out_dir):
        path = os.path.join(out_dir, self.filename)
        logging.debug(path)

        try:
            with urllib.request.urlopen(self.url) as response, open(path, 'wb+') as out_file:
                shutil.copyfileobj(response, out_file)
        except urllib.error.HTTPError as e:
            logging.exception(e)

    @staticmethod
    def from_json(data) -> 'Post':
        return Post(data['id'], bool(data['is_self']), Url(data['url']))


def subreddit_to_url(subreddit: str) -> Url:
    return Url('https://www.reddit.com/r/{}'.format(subreddit))


def fetch_posts(subreddit: Url, limit=25, after=''):
    params = urllib.parse.urlencode({'limit': limit, 'after': after})
    url = '{}.json?{}'.format(subreddit, params)
    req = urllib.request.Request(url, headers={'User-Agent': USER_AGENT})
    resp = urllib.request.urlopen(req)
    result = json.loads(resp.read().decode('utf-8'))
    return [Post.from_json(p['data']) for p in result['data']['children']]


def download_images(subreddit_name, n, out_dir) -> None:
    posts = fetch_posts(subreddit_name, limit=n)
    num_images = 0
    while True:
        for post in posts:
            if post.is_valid:
                post.download(out_dir)
                num_images += 1
                if num_images >= n:
                    return

        posts = fetch_posts(subreddit_name, limit=n, after=posts[-1].id)


def download_all_images(subreddits: typing.Sequence[Url], n, out_dir):
    """Download n images in parallel for each subreddit to out_dir."""
    with multiprocessing.Pool() as pool:
        for subreddit in subreddits:
            pool.apply(download_images, args=(subreddit, n, out_dir))


def parse_cli():
    parser = argparse.ArgumentParser(description='Download images from Reddit.')
    parser.add_argument('--subreddits', nargs='*')
    parser.add_argument('--multireddit', nargs='?', help='USER/multi_name')
    parser.add_argument('--out', required=True)
    parser.add_argument('-v', '--verbose', dest='verbose', action='store_true')
    return parser.parse_args()


def main():
    args = parse_cli()
    logging_level = logging.DEBUG if args.verbose else logging.INFO
    logging.basicConfig(stream=sys.stdout, level=logging_level)

    if not args.subreddits and not args.multireddit:
        subreddits = [subreddit_to_url(s) for s in DEFAULT_SUBREDDITS]
    else:
        subreddits = []
        if args.subreddits:
            subreddits.extend([subreddit_to_url(s) for s in args.subreddits])

        if args.multireddit:
            user, multi = args.multireddit.split('/')
            multi_url = 'https://www.reddit.com/user/{}/m/{}'.format(user, multi)
            subreddits.append(multi_url)

    download_all_images(subreddits, IMAGES_PER_SUBREDDIT, args.out)

if __name__ == '__main__':
    main()
